### PR TITLE
Add autoformatting to ts bindings & navigator frontend

### PR DESCRIPTION
**Pull request by realvictorprm**
Includes complete reformat of our webdev files (js, tsx, html etc.). Sadly huge changes because there was a lot to format.

changelog_begin
changelog_end

### Pull Request Checklist

- [x] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [x] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [x] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description
- [ ] If you mean to change the status of a component, please make sure you keep [the Component Status page](https://github.com/digital-asset/daml/blob/main/docs/source/support/component-statuses.rst) up to date.

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: .prettierrc.yaml [hunk 1]
```diff
@@ -0,0 +1,4 @@
+trailingComma: "es5"
+semi: false
```
Comment (realvictorprm) [inline .prettierrc.yaml:2]: Maybe I'm wrong here but I consider using semicolons everywhere as a relict of the past.

File: language-support/ts/codegen/README.md [hunk 2]
```diff
@@ -322,31 +369,32 @@ daml.registerTemplate(Iou);
...
+- Companion objects for those types;
+- An `Iou` template associated type definition `Key` in the `Iou` namespace;
+- A module scoped function call to `registerTemplate` to add the `Iou` object to a central registry.
 
 #### Daml Interfaces
 
 An `Asset` interface.
 
-```haskell
+````haskell
```
Comment (realvictorprm) [inline language-support/ts/codegen/README.md:383]: looks like some readme changes are wrong. At least this one looks wrong. Maybe I should exclude the readme files from formatting (wasn't my intention to have them formatted too but from what I saw all looked fine. First fault I found so far).

File: fmt.sh [hunk 3]
```diff
@@ -143,6 +143,9 @@ else
   run javafmt "${javafmt_args[@]:-}" ${java_files[@]:-}
 fi
 
+# check for typescript code style
+run prettier --write language-support/ts navigator/frontend
```
Comment (stefanobaghino-da) [inline fmt.sh:147]: This doesn't just check, but actually rewrites your files ([documentation](https://prettier.io/docs/en/cli.html)).

Formatting TS code should honor the same conditions as other language-specific formatters in this script (Scala, Haskell, Java, Starlark):
- if the script is invoked with `--test`, this should only check and not rewrite (CI uses this)
- if the script is invoked with `--diff`, either the check or the rewrite should only look at the changes since the current branch split away from `main` (to minimize the cost of formatting locally)

File: language-support/ts/daml-ledger/README.md [hunk 4]
```diff
@@ -33,40 +33,38 @@ Please consult the
 [documentation](https://docs.daml.com/app-dev/bindings-ts/daml-ledger/classes/_index_.ledger.html)
 for their exact signatures.
 
-`create`
---------
+## `create`
```
Comment (stefanobaghino-da) [inline language-support/ts/daml-ledger/README.md:36]: It looks like this was previously formatted as an RST document. This PR is already huge and if you have further work that you have done while addressing the formatting it should go in a separate PR to give it the proper attention (in a PR that formats a lot of code, this can easily go unnoticed).

File: language-support/ts/daml-react/defaultLedgerContext.ts [hunk 5]
```diff
@@ -82,8 +109,12 @@ export function useFetch<T extends object, K, I extends string>(template: Templa
  *
  * @return The fetched contract.
  */
-export function useFetchByKey<T extends object, K, I extends string>(template: Template<T, K, I>, keyFactory: () => K, keyDeps: readonly unknown[]): FetchResult<T, K, I> {
-  return ledgerContext.useFetchByKey(template, keyFactory, keyDeps);
+export function useFetchByKey<T extends object, K, I extends string>(
+  template: Template<T, K, I>,
+  keyFactory: () => K,
+  keyDeps: readonly unknown[]
```
Comment (stefanobaghino-da) [inline language-support/ts/daml-react/defaultLedgerContext.ts:115]: Can `prettier` be configured to had dangling commas in parameter lists? I found them to be a nice way to reduce diff noise in Scala (I'm not even sure whether that's feasible in TS itself, just asking).

File: language-support/ts/daml-ledger/README.md (hunk 4 above)
Comment (stefanobaghino-da) [reply language-support/ts/daml-ledger/README.md:36]: I just noticed that this happened to all MarkDown files, so I assume this is performed automatically by `prettier`, so nothing contrary to merging those as part of this PR.

File: fmt.sh (hunk 3 above)
Comment (stefanobaghino-da) [reply fmt.sh:147]: Any reason not to extend the check to the entire repository?

**Review by stefanobaghino-da — DISMISSED** [review]
Thanks for working on this. I very much appreciate automating away formatting from engineers. I appreciate the uniformity of style.

One important point to address here: have `prettier` honor `--diff` and `--test` like all other checks. This can be achieved by either mapping these options onto the underlying formatter (as in the case of Scala) or building the functionality yourself (as in the case of Haskell and Java).

File: language-support/ts/daml-react/defaultLedgerContext.ts (hunk 5 above)
Comment (realvictorprm) [reply language-support/ts/daml-react/defaultLedgerContext.ts:115]: I think this is possible but need to check.

File: fmt.sh (hunk 3 above)
Comment (realvictorprm) [reply fmt.sh:147]: Performance of prettier is too bad for that, at least so far I've tested.

Comment (SamirTalwar-DA) [reply fmt.sh:147]: I expect `prettier` will be fast enough if you limit it to files that are seen by `git`, e.g. by running `git ls-files | xargs prettier`.

Comment (realvictorprm) [reply fmt.sh:147]: That's a good idea, thank you I will try that!

File: language-support/ts/daml-react/defaultLedgerContext.ts (hunk 5 above)
Comment (SamirTalwar-DA) [reply language-support/ts/daml-react/defaultLedgerContext.ts:115]: Create a _.prettierrc_ file as follows:

```
trailingComma: all
```

[Docs.](https://prettier.io/docs/en/options.html)

File: docs/theme/bower_components_static/bourbon/readme.md [hunk 6]
```diff
@@ -1,12 +1,12 @@
 [![Bourbon Sass Mixin Library](http://bourbon.io/images/shared/bourbon-logo.png)](http://bourbon.io)
```
Comment (SamirTalwar-DA) [inline docs/theme/bower_components_static/bourbon/readme.md:1]: Do we own this component? Or is this file copied from somewhere?

File: docs/theme/bower_components_static/bourbon/.bower.json [hunk 7]
```diff
@@ -23,4 +23,4 @@
   "_source": "https://github.com/thoughtbot/bourbon.git",
   "_target": "4.0.2",
   "_originalSource": "bourbon"
-}
\ No newline at end of file
+}
```
Comment (SamirTalwar-DA) [inline docs/theme/bower_components_static/bourbon/.bower.json:26]: Not convinced we should be auto-formatting these files; they look like they were copied from somewhere.

File: docs/theme/js/nice-select.min.js [hunk 8]
```diff
@@ -4,4 +4,119 @@
 /*  jQuery Nice Select - v1.0
```
Comment (SamirTalwar-DA) [inline docs/theme/js/nice-select.min.js:4]: OK, now I'm convinced; don't auto-format minified code. I think you can exclude _docs/theme_ (though I also think a lot of this stuff should not be checked in.)

File: fmt.sh [hunk 9]
```diff
@@ -20,6 +20,9 @@ dade_copyright_arg=update
 buildifier_target=//:buildifier-fix
 security_update_args=()
 
+## Constants ##
+webdef_files_regex="\.(ts|js|md|html|json|tsx)$"
```
Comment (SamirTalwar-DA) [inline fmt.sh:24]: Sort logically, please. Either alphabetically or grouping related extensions (e.g. `md|html|json|js|ts|tsx`).

File: fmt.sh [hunk 10]
```diff
@@ -133,6 +136,7 @@ run dade-copyright-headers "$dade_copyright_arg" .
 if [ "$diff_mode" = "true" ]; then
```
Comment (SamirTalwar-DA) [inline fmt.sh:136]: Whether we are diffing or not does not imply whether we are testing or not.

I suggest setting `prettier_args` to `('--write')` or `('--check')` appropriately above, and using it below.

**Review by SamirTalwar-DA — COMMENTED** [review]
Good start, but I think you have taken too much on at once.

I recommend doing this in smaller chunks. Start with Markdown, then HTML, then JSON, then JavaScript and TypeScript.

File: .prettierrc.yaml [hunk 11]
```diff
@@ -0,0 +1,7 @@
+# Copyright (c) 2022 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
```
[the remark this answers is not in the document]
Comment (stefanobaghino-da) [reply .prettierrc.yaml:1]: Nice catch.

File: fmt.sh (hunk 3 above)
Comment (stefanobaghino-da) [reply fmt.sh]: This should be the behavior of `--diff`. It should run on the whole repo otherwise. CI runs with `--test` only.

Comment (SamirTalwar-DA) [reply fmt.sh]: `git ls-files` lists all files in the repository except un-indexed or ignored files.

File: .prettierrc.yaml (hunk 11 above)
[the remark this answers is not in the document]
Comment (SamirTalwar-DA) [reply .prettierrc.yaml:1]: We may also want to switch to single quotes if it makes the diff smaller.

[the remark this answers is not in the document]
Comment (realvictorprm) [reply .prettierrc.yaml:1]: I'm usually all in for reducing the diff, however in this case I much prefer to aim for a good code style that is well readable because whatever we do the diff will nevertheless be huge.

File: dev-env/windows/manifests/curl.json [hunk 12]
```diff
@@ -1,39 +1,39 @@
 {
```
[the remark this answers is not in the document]
Comment (realvictorprm) [reply dev-env/windows/manifests/curl.json:1]: I'll add it to the ignore file.

File: docs/theme/js/nice-select.min.js (hunk 8 above)
Comment (realvictorprm) [reply docs/theme/js/nice-select.min.js:4]: OK that convinced me too.

File: .prettierrc.yaml (hunk 11 above)
[the remark this answers is not in the document]
Comment (SamirTalwar-DA) [reply .prettierrc.yaml:1]: If you feel like having that debate on what's "readable", please go ahead. :stuck_out_tongue:

[the remark this answers is not in the document]
Comment (garyverhaegen-da) [reply .prettierrc.yaml:1]: AFAICT the only difference between single and double quotes in JS strings is which quotes you need to escape if you want them literally in your string. Based on that:

- Apostrophes are much more common than double quotes in English, which would point in favour of double quotes for JS strings.
- JSON only allows double quotes, so using that in JS makes it easier to copy/paste between the two — a niche use-case to be sure, but still a stronger argument than "I like TYPE_OF_QUOTE better".

File: .prettierignore [hunk 13]
```diff
@@ -0,0 +1,11 @@
+# Autogenerated json files, which are not intended to be read by humans (and which shouldn't be formatted either).
+maven_install.json
+maven_install_2.13.json
+ghcide_snapshot.json
+stackage_snapshot_windows.json
+stackage_snapshot.json
+dev-env/windows/manifests/curl.json
```
Comment (SamirTalwar-DA) [inline .prettierignore:7]:
```suggestion
dev-env/windows/manifests/*.json
```

File: .prettierrc.yaml (hunk 11 above)
[the remark this answers is not in the document]
Comment (SamirTalwar-DA) [reply .prettierrc.yaml:1]: I also prefer double-quotes; I would just like to move that conversation to a separate PR.

File: navigator/frontend/src/applets/about/index.tsx [hunk 14]
```diff
@@ -112,60 +127,63 @@ const VSpace = styled.div`
...
         <VSpace />
         <div>
-        <p>
-          Copyright Notice
-          Copyright (c) 2022 Digital Asset (Switzerland) GmbH and/or its affiliates.
-          All rights reserved.
-        </p>
+          <p>
+            Copyright Notice Copyright (c) 2022 Digital Asset (Switzerland) GmbH
+            and/or its affiliates. All rights reserved.
```
Comment (SamirTalwar-DA) [inline navigator/frontend/src/applets/about/index.tsx:180]: I think line breaks were intended here. Might be worth verifying.

File: .prettierrc.yaml (hunk 11 above)
[the remark this answers is not in the document]
Comment (stefanobaghino-da) [reply .prettierrc.yaml:1]: Seems like double quotes have a good consensus behind them.

File: .prettierrc.yaml (hunk 1 above)
Comment (realvictorprm) [reply .prettierrc.yaml:5]: OK so overall I would still like to get rid of the semicolons because I'm in the camp of people who prefer to reduce the noise in code as much as possible and semicolons are considered by myself as noise during visual parsing.

However after investigating further it probably is best to ask the autoformatter to add semicolons while we can still code mixed style locally. I will change the option then.

Comment (garyverhaegen-da) [reply .prettierrc.yaml:5]: In sane languages semicolons are either mandatory or decorative, and thus can count as noise.

In JavaScript, semicolons are **optional but semantically significant**, so you're going to be much, much better off always writing them down and acting as though they were mandatory.

Comment (realvictorprm) [reply .prettierrc.yaml]: To me the mechanism which automatically adds the semicolons is actually only an adjusted parser which follows parsing rules that don't require semicolons. So overall it is just necessary to understand why a specific statement is different to one without semicolons. In general it is good if one learns about these, especially because they are not that difficult (happily referring to a stackoverflow [answer](https://stackoverflow.com/a/48089473) about this topic).

File: .prettierrc.yaml (hunk 11 above)
[the remark this answers is not in the document]
Comment (realvictorprm) [reply .prettierrc.yaml:1]: I disagree about moving to a different PR, let's please only do the formatting once and right

File: .prettierrc.yaml (hunk 1 above)
Comment (garyverhaegen-da) [reply .prettierrc.yaml]: I don't understand how
```javascript
a = b + c
(d + e).foo()
```
is not a very strong argument for always having semicolons.

Comment (realvictorprm) [reply .prettierrc.yaml]: It is, but knowing it is also good because it is an important detail of how fecked up javascript is after all. I mean I did decide in favour of semicolons after all because of this new information. However this still won't stop me to not use semicolons all the time while doing local development, I have a formatter which adds it after all :smile:

File: fmt.sh [hunk 15]
```diff
@@ -44,7 +45,7 @@ check_diff() {
   # $1 merge_base
   # $2 regex
   # "${@:3}" command
-  changed_files=$(git diff --name-only --diff-filter=ACMRT "$1" | grep $2 || [[ $? == 1 ]])
+  changed_files=$(git diff --name-only --diff-filter=ACMRT "$1" | grep -E $2 || [[ $? == 1 ]])
```
Comment (stefanobaghino-da) [inline fmt.sh:48]: Using extended regular expressions seems substantial. Why the change and what's the impact on other formatter that use the information here?

File: fmt.sh [hunk 16]
```diff
@@ -133,6 +140,7 @@ run dade-copyright-headers "$dade_copyright_arg" .
 if [ "$diff_mode" = "true" ]; then
   check_diff $merge_base '\.hs$' hlint -j4
   check_diff $merge_base '\.java$' javafmt "${javafmt_args[@]:-}"
+  check_diff $merge_base '\.(ts|tsx)$' npx pprettier --check
```
Comment (stefanobaghino-da) [inline fmt.sh:143]: Is the name `pprettier` a typo? I see it appearing again, so not sure. Furthermore, shouldn't you use `prettier_args` here?

File: fmt.sh [hunk 17]
```diff
@@ -88,6 +89,12 @@ USAGE
   esac
 done
 
+if [[ $is_test = 1 ]]; then
+  prettier_args="--check"
+else
+  prettier_args="--write"
+fi
+
```
Comment (stefanobaghino-da) [inline fmt.sh:97]: Move this up where all other `--test` check are run, we want to keep them together so that we can act on all of them at once if needed. Look at how this is done currently and replicate the same approach.

File: fmt.sh (hunk 15 above)
Comment (realvictorprm) [reply fmt.sh:48]: I think this can be removed

**Comment by realvictorprm** [comment]
OK, all addressed. Feel free to take another look.

**Comment by stefanobaghino-da** [comment]
> OK, all addressed. Feel free to take another look.

I just checked out this branch, removed a semi-colon from `language-support/ts/codegen/tests/ts/build-and-lint-test/src/index.ts` and ran `./fmt.sh --test --diff`. I would expect the script to only touch a single file and return a non-zero value. Instead, the following happens

```
$ npx pprettier --check compatibility/bazel_tools/create-daml-app/index.test.ts compiler/daml-extension/src/extension.ts compiler/daml-extension/test/extension.test.ts compiler/daml-extension/test/index.ts language-support/ts/codegen/tests/ts/build-and-lint-test/src/__tests__/test.ts language-support/ts/codegen/tests/ts/build-and-lint-test/src/index.ts language-support/ts/daml-ledger/index.test.ts language-support/ts/daml-ledger/index.ts language-support/ts/daml-react/createLedgerContext.ts language-support/ts/daml-react/defaultLedgerContext.ts language-support/ts/daml-react/index.test.ts language-support/ts/daml-react/index.ts language-support/ts/daml-types/index.test.ts language-support/ts/daml-types/index.ts navigator/frontend/declarations.d.ts navigator/frontend/src/api/Queries.ts navigator/frontend/src/applets/about/index.tsx navigator/frontend/src/applets/app/index.tsx navigator/frontend/src/applets/configsource/index.tsx navigator/frontend/src/applets/contract/ContractComponent.tsx navigator/frontend/src/applets/contract/Exercise.tsx navigator/frontend/src/applets/contract/index.tsx navigator/frontend/src/applets/contracts/columns.tsx navigator/frontend/src/applets/contracts/data.tsx navigator/frontend/src/applets/contracts/index.tsx navigator/frontend/src/applets/customview/index.tsx navigator/frontend/src/applets/info/index.tsx navigator/frontend/src/applets/page/index.tsx navigator/frontend/src/applets/template/TemplateComponent.tsx navigator/frontend/src/applets/template/index.tsx navigator/frontend/src/applets/templatecontracts/columns.tsx navigator/frontend/src/applets/templatecontracts/data.tsx navigator/frontend/src/applets/templatecontracts/index.tsx navigator/frontend/src/applets/templates/columns.tsx navigator/frontend/src/applets/templates/data.tsx navigator/frontend/src/applets/templates/index.tsx navigator/frontend/src/browsercheck.ts navigator/frontend/src/components/Frame.tsx navigator/frontend/src/components/Icon.tsx navigator/frontend/src/components/Link.tsx navigator/frontend/src/components/Navbar.tsx navigator/frontend/src/config/api/v1-load.ts navigator/frontend/src/config/api/v1.ts navigator/frontend/src/config/api/v2-load.ts navigator/frontend/src/config/api/v2.ts navigator/frontend/src/config/either.ts navigator/frontend/src/config/index.ts navigator/frontend/src/config/utils.ts navigator/frontend/src/declarations.d.ts navigator/frontend/src/index.tsx navigator/frontend/src/polyfills.ts navigator/frontend/src/redux-router/src/index.ts navigator/frontend/src/routes/index.ts navigator/frontend/src/theme.ts navigator/frontend/src/ui-core/declarations.d.ts navigator/frontend/src/ui-core/src/AdvanceTime/index.tsx navigator/frontend/src/ui-core/src/ArgumentDisplay/Guide.tsx navigator/frontend/src/ui-core/src/ArgumentDisplay/index.tsx navigator/frontend/src/ui-core/src/Autosuggest/Guide.tsx navigator/frontend/src/ui-core/src/Autosuggest/index.tsx navigator/frontend/src/ui-core/src/Breadcrumbs/Guide.tsx navigator/frontend/src/ui-core/src/Breadcrumbs/index.tsx navigator/frontend/src/ui-core/src/Button/Guide.tsx navigator/frontend/src/ui-core/src/Button/index.tsx navigator/frontend/src/ui-core/src/ChoicesButton/Guide.tsx navigator/frontend/src/ui-core/src/ChoicesButton/index.tsx navigator/frontend/src/ui-core/src/ContractTable/ApolloDataProvider.ts navigator/frontend/src/ui-core/src/ContractTable/Guide.tsx navigator/frontend/src/ui-core/src/ContractTable/index.tsx navigator/frontend/src/ui-core/src/DataTable/Guide.tsx navigator/frontend/src/ui-core/src/DataTable/index.tsx navigator/frontend/src/ui-core/src/DateTimePicker/calendar.tsx navigator/frontend/src/ui-core/src/DateTimePicker/index.tsx navigator/frontend/src/ui-core/src/DateTimePicker/timeInput.tsx navigator/frontend/src/ui-core/src/Frame/index.tsx navigator/frontend/src/ui-core/src/Guide/index.tsx navigator/frontend/src/ui-core/src/Icon/Guide.tsx navigator/frontend/src/ui-core/src/Icon/index.tsx navigator/frontend/src/ui-core/src/Input/index.tsx navigator/frontend/src/ui-core/src/Label/index.tsx navigator/frontend/src/ui-core/src/Link/Guide.tsx navigator/frontend/src/ui-core/src/Link/index.tsx navigator/frontend/src/ui-core/src/Modal/Guide.tsx navigator/frontend/src/ui-core/src/Modal/index.tsx navigator/frontend/src/ui-core/src/NavBar/index.tsx navigator/frontend/src/ui-core/src/NestedForm/index.tsx navigator/frontend/src/ui-core/src/ParameterForm/ApolloDataProvider.ts navigator/frontend/src/ui-core/src/ParameterForm/ContractIdInput.tsx navigator/frontend/src/ui-core/src/ParameterForm/Guide.tsx navigator/frontend/src/ui-core/src/ParameterForm/index.tsx navigator/frontend/src/ui-core/src/Popover/Guide.tsx navigator/frontend/src/ui-core/src/Popover/index.tsx navigator/frontend/src/ui-core/src/RouteMatcher/index.ts navigator/frontend/src/ui-core/src/SearchInput/Guide.tsx navigator/frontend/src/ui-core/src/SearchInput/index.tsx navigator/frontend/src/ui-core/src/Select/Guide.tsx navigator/frontend/src/ui-core/src/Select/index.tsx navigator/frontend/src/ui-core/src/SidebarLink/Guide.tsx navigator/frontend/src/ui-core/src/SidebarLink/index.tsx navigator/frontend/src/ui-core/src/Strong/Guide.tsx navigator/frontend/src/ui-core/src/Strong/index.tsx navigator/frontend/src/ui-core/src/Table/HeaderCell.tsx navigator/frontend/src/ui-core/src/Table/index.tsx navigator/frontend/src/ui-core/src/TableActionBar/Guide.tsx navigator/frontend/src/ui-core/src/TableActionBar/TabLink.tsx navigator/frontend/src/ui-core/src/TableActionBar/index.tsx navigator/frontend/src/ui-core/src/TimeInput/Guide.tsx navigator/frontend/src/ui-core/src/TimeInput/index.tsx navigator/frontend/src/ui-core/src/Tooltip/Guide.tsx navigator/frontend/src/ui-core/src/Tooltip/index.tsx navigator/frontend/src/ui-core/src/Truncate/Guide.tsx navigator/frontend/src/ui-core/src/Truncate/index.tsx navigator/frontend/src/ui-core/src/api/DamlLfType.ts navigator/frontend/src/ui-core/src/api/DamlLfValue.ts navigator/frontend/src/ui-core/src/api/OpaqueTypes.ts navigator/frontend/src/ui-core/src/api/Queries.ts navigator/frontend/src/ui-core/src/devel/index.tsx navigator/frontend/src/ui-core/src/index.ts navigator/frontend/src/ui-core/src/ledger-watcher/UI.tsx navigator/frontend/src/ui-core/src/ledger-watcher/icons.tsx navigator/frontend/src/ui-core/src/ledger-watcher/index.ts navigator/frontend/src/ui-core/src/session/UI.tsx navigator/frontend/src/ui-core/src/session/actions.ts navigator/frontend/src/ui-core/src/session/index.ts navigator/frontend/src/ui-core/src/session/reduce.ts navigator/frontend/src/ui-core/src/session/types.ts navigator/frontend/src/ui-core/src/theme.ts navigator/frontend/src/ui-core/src/types.ts navigator/frontend/src/ui-core/src/util.tsx navigator/frontend/src/ui-core/src/withLedgerTime/index.tsx templates/create-daml-app-test-resources/index.test.ts templates/create-daml-app/ui/src/Credentials.ts templates/create-daml-app/ui/src/components/App.test.tsx templates/create-daml-app/ui/src/components/App.tsx templates/create-daml-app/ui/src/components/MainScreen.tsx templates/create-daml-app/ui/src/components/PartyListEdit.tsx templates/create-daml-app/ui/src/index.tsx
npx: installed 46 in 2.876s
The option "check" is unknown. Did you mean the following one?
-h, --help  Output usage information
```

- the command touches a lot of files, not just one.
- the command itself seems to not recognize `check` as an option
- the command still returns as no error occurred
  - I want the formatting check to return with an error if any file does not comply

You may want to run a few tests yourself to ensure that the script behaves as it should.

I'm also not sure that the `write` option is valid either for `pprettier`. From a quick look at `npx pprettier --help` neither appear in the help.

Is `npx` (whatever it is) the only option to run this? It seems to re-run some form of installation of something every time it runs and it would be preferable not to add this time, the idea is that `./fmt.sh --diff --check` should be as quick as possible to serve as a pre-push hook. It's already pretty slow, no reason to slow it down further.

**Comment by realvictorprm** [comment]
Darn, many thanks for the feedback. I didn't think about trying to combine
both flags 😅  besides that I just noticed what goes wrong about the part
with checking only changed files 🤦‍♂️ will fix that tomorrow. Sorry for
the inconvenience

Stefano Baghino ***@***.***> schrieb am Mi., 9. Feb. 2022,
20:35:

**Comment by stefanobaghino-da** [comment]
Regarding the usage of `npx`, [it looks like you can install `prettier` via `nix`](https://github.com/NixOS/nixpkgs/issues/126804#issuecomment-860921762), which could probably solve the issue of the long startup time.

**Comment by cocreature** [comment]
> Regarding the usage of `npx`, [it looks like you can install `prettier` via `nix`](https://github.com/NixOS/nixpkgs/issues/126804#issuecomment-860921762), which could probably solve the issue of the long startup time.

prettier yes, parallel-prettier not quite as easily (but not quite as hard either). If you want to stick with npm, invoke it via bazel and you’ll get caching making this much less of an issue than with npx.

**Comment by stefanobaghino-da** [comment]
> > Regarding the usage of `npx`, [it looks like you can install `prettier` via `nix`](https://github.com/NixOS/nixpkgs/issues/126804#issuecomment-860921762), which could probably solve the issue of the long startup time.
> 
> prettier yes, parallel-prettier not quite as easily (but not quite as hard either). If you want to stick with npm, invoke it via bazel and you’ll get caching making this much less of an issue than with npx.

`pprettier` seems incompatible with `prettier`'s CLI options (quite confusingly) so it's not a drop-in replacement and requires some work to be adapted apparently. How much time does it take to process a file? We have 137 `.ts` and `.tsx` files (116 of which in Navigator) and I would expect a relatively large PR to maybe touch a dozen of those that one may want to check locally with `--diff` enabled.

**Comment by cocreature** [comment]
if prettier is fast enough, I’m all for it

**Comment by realvictorprm** [comment]
For the case of just calling `fmt.sh` without any args pprettier is a lot faster than prettier. This is why I swapped out prettier with pprettier. Besides, I'm surprised that all ts files are picked up in the case of `--test --diff`. Moreover it is weird that you run into that error because I do not run into that error locally. very weird :( I'm still investigating what is going on.

**Comment by stefanobaghino-da** [comment]
> OK so far I see `--diff` resolves to
> 
> ```
> export merge_base="$(git merge-base origin/main HEAD)"
> git diff --name-only --diff-filter=ACMRT "$merge_base"
> ```
> 
> which lists all the files which were changed in respect to the main branch thus being the source for listing all typescript files. So this isn't caused by my code, it is existing behaviour.

I'm not sure you're going to the root cause. The Java formatter is clearly behaving in the correct way.

**Comment by stefanobaghino-da** [comment]
> For the case of just calling `fmt.sh` without any args pprettier is a lot faster than prettier. This is why I swapped out prettier with pprettier.

If it's slower by a few seconds on CI it's fine, I'm mostly worried about running it locally.

**Comment by stefanobaghino-da** [comment]
Ok, I got it. It's trying to run on all files precisely because I'm on your branch, and your branch touched all the files, so `--diff` returns all the files in the branch. Sorry for the fuss. Still, I don't understand why this doesn't show up on your machine.

**Review by stefanobaghino-da — APPROVED** [review]
I tested everything again, all good, thanks.

**Comment by realvictorprm** [comment]
> Ok, I got it. It's trying to run on all files precisely because I'm on your branch, and your branch touched all the files, so `--diff` returns all the files in the branch. Sorry for the fuss. Still, I don't understand why this doesn't show up on your machine.

I can replicate the behaviour of the files but not that the part where you called pprettier via npx where it complained about `--check` being unknown. 

Anyways, glad that all is now resolved :) Many thanks for testing it on your machine too!

**Comment by realvictorprm** [comment]
now finally I think everything should work "perfectly :tm:"
